File: app/src/main/java/io/inabsentia/superhangman/adapter/HighScoreAdapter.java
package io.inabsentia.superhangman.adapter;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import java.util.List;

import io.inabsentia.superhangman.R;
import io.inabsentia.superhangman.item.HighScoreItem;
import io.inabsentia.superhangman.singleton.App;

public final class HighScoreAdapter extends ArrayAdapter<HighScoreItem> {

    private List<HighScoreItem> items;
    private Context mContext;
    private App app = App.getInstance();

    public HighScoreAdapter(@NonNull Context context, int resource, @NonNull List<HighScoreItem> items) {
        super(context, resource, items);
        this.items = items;
        this.mContext = context;
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        View mConvertView = convertView;
        HighScoreItem item = getItem(position);
        HighScoreItemViewHolder viewholder;

        if (mConvertView == null) {
            LayoutInflater inflater = LayoutInflater.from(mContext);
            mConvertView = inflater.inflate(R.layout.high_score_item, parent, false);
            viewholder = new HighScoreItemViewHolder(mConvertView);
            mConvertView.setTag(viewholder);
        } else {
            viewholder = (HighScoreItemViewHolder) mConvertView.getTag();
        }

        if (item != null) {
            viewholder.tvName.setText(item.getUsername());
            viewholder.tvScore.setText(item.getHighscore());
        }

        return mConvertView;
    }

    private static class HighScoreItemViewHolder extends RecyclerView.ViewHolder {
        private TextView tvName, tvScore;

        public HighScoreItemViewHolder(View view) {
            super(view);
            tvName = view.findViewById(R.id.username_field);
            tvScore = view.findViewById(R.id.score_field);
        }
    }

}

File: app/src/main/java/io/inabsentia/superhangman/adapter/HighScoreAdapter.kt
package io.inabsentia.superhangman.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.widget.ArrayAdapter
import android.widget.TextView
import io.inabsentia.superhangman.R
import io.inabsentia.superhangman.data.dto.HighScoreDTO

class HighScoreAdapter(private val highScoreList: List<HighScoreDTO>, private val mContext: Context) : ArrayAdapter<HighScoreDTO>(mContext, R.layout.high_score_item, highScoreList) {

    private var lastPosition = -1

    /*
     * ViewHolder that encapsulates the objects
     * that are used in the list view.
     */
    private class ViewHolder {
        internal var tvPos: TextView? = null
        internal var tvNameLabel: TextView? = null
        internal var tvName: TextView? = null
        internal var tvHighScoreLabel: TextView? = null
        internal var tvHighScore: TextView? = null
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var mConvertView = convertView

        val view: View
        val viewHolder: ViewHolder

        val highScoreDTO = highScoreList[position]

        if (mConvertView == null) {
            viewHolder = ViewHolder()
            val inflater = LayoutInflater.from(context)
            mConvertView = inflater.inflate(R.layout.high_score_item, parent, false)

            viewHolder.tvPos = mConvertView.findViewById(R.id.position_view)
            viewHolder.tvNameLabel = mConvertView.findViewById(R.id.high_score_user_name_label)
            viewHolder.tvName = mConvertView.findViewById(R.id.high_score_user_name)
            viewHolder.tvHighScoreLabel = mConvertView.findViewById(R.id.high_score_user_name_label)
            viewHolder.tvHighScore = mConvertView.findViewById(R.id.high_score)

            view = mConvertView
            mConvertView.tag = viewHolder
        } else {
            viewHolder = mConvertView.tag as ViewHolder
            view = mConvertView
        }

        val animation = AnimationUtils.loadAnimation(mContext, if (position > lastPosition) R.anim.up_from_bottom else R.anim.down_from_top)
        view.startAnimation(animation)
        lastPosition = position

        viewHolder.tvPos!!.text = (position + 1).toString()
        viewHolder.tvName!!.text = highScoreDTO.name
        viewHolder.tvHighScore!!.text = highScoreDTO.highscore.toString()

        return mConvertView!!
    }

}